File: go/src/514.自由之路.go
/*
 * @lc app=leetcode.cn id=514 lang=golang
 *
 * [514] 自由之路
 */

// @lc code=start
func findRotateSteps(ring string, key string) int {
	pos := make(map[byte][]int) // ring中字符的位置
	for i := 0; i < len(ring); i++ {
		pos[ring[i]] = append(pos[ring[i]], i)
	}

	cache := make([][]int, len(ring))
	for i := 0; i < len(ring); i++ {
		cache[i] = make([]int, len(key))
	}

	return len(key) + dfs(ring, key, 0, 0, pos, cache)
}

func dfs(ring, key string, rindex, kindex int, pos map[byte][]int, cache [][]int) int {
	if kindex == len(key) {
		return 0
	}

	if cache[rindex][kindex] > 0 {
		return cache[rindex][kindex]
	}

	choiceNum, min := len(pos[key[kindex]]), math.MaxInt32
	for i := 0; i < choiceNum; i++ {
		p := pos[key[kindex]][i]
		cost := abs(rindex - p)
		if len(ring) - cost < cost {
			cost = len(ring) - cost
		}

		tmp := dfs(ring, key, p, kindex + 1, pos, cache)
		if tmp + cost  < min {
			min = tmp + cost
		}
	}
	cache[rindex][kindex] = min

	return cache[rindex][kindex]
}

func abs(n int) int {
	y := n >> 31
	return (n ^ y) - y
}
// @lc code=end

File: rust/src/514.自由之路.rs
/*
 * @lc app=leetcode.cn id=514 lang=rust
 *
 * [514] 自由之路
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn find_rotate_steps(ring: String, key: String) -> i32 {
        if ring.is_empty() || key.is_empty() {
            return 0;
        }
        let n = ring.len();
        let m = key.len();
        const MAX: usize = std::i32::MAX as usize;
        let mut f = vec![vec![MAX; n]; 2];
        let mut pos = HashMap::new();
        for (i, &c) in ring.as_bytes().iter().enumerate() {
            pos.entry(c).or_insert(vec![]).push(i);
        }
        let rs = ring.as_bytes();
        let ks = key.as_bytes();
        if let Some(arr) = pos.get(&ks[0]) {
            for &i in arr {
                f[1][i] = Self::calc(0, i, n) + 1;
            }
        }
        let mut idx = 0;
        for i in 1..m {
            let nidx = idx ^ 1;
            for j in 0..n {
                f[idx][j] = MAX;
                if rs[j] != ks[i] {
                    continue;
                }
                if let Some(pre) = pos.get(&ks[i - 1]) {
                    for &k in pre {
                        f[idx][j] = f[idx][j].min(f[nidx][k] + Self::calc(j, k, n) + 1);
                    }
                }
            }
            idx = nidx;
        }
        let mut ans = std::i32::MAX as usize;
        for &v in f[idx ^ 1].iter() {
            ans = ans.min(v);
        }
        ans as i32
    }

    fn calc(a: usize, b: usize, n: usize) -> usize {
        if a < b {
            (b - a).min(n - b + a)
        } else {
            (a - b).min(n - a + b)
        }
    }
}
// @lc code=end
